contracts/exchangeIssuance/ExchangeIssuanceZeroEx.sol

```diff
@@ -0,0 +1,561 @@
+
+    address public immutable WETH;
+    mapping(address => IssuanceModuleData) public allowedIssuanceModules;
+
+    IController public immutable setController;
+
+    address public swapTarget;
+
+    /* ============ Events ============ */
+
+    event ExchangeIssue(
+        address indexed _recipient,     // The recipient address of the issued SetTokens
+        ISetToken indexed _setToken,    // The issued SetToken
+        IERC20 indexed _inputToken,     // The address of the input asset(ERC20/ETH) used to issue the SetTokens
+        uint256 _amountInputToken,      // The amount of input tokens used for issuance
+        uint256 _amountSetIssued        // The amount of SetTokens received by the recipient
+    );
+
+    event ExchangeRedeem(
+        address indexed _recipient,     // The recipient address which redeemed the SetTokens
+        ISetToken indexed _setToken,    // The redeemed SetToken
+        IERC20 indexed _outputToken,    // The address of output asset(ERC20/ETH) received by the recipient
+        uint256 _amountSetRedeemed,     // The amount of SetTokens redeemed for output tokens
+        uint256 _amountOutputToken      // The amount of output tokens received by the recipient
+    );
+
+    /* ============ Modifiers ============ */
+
+    modifier isWhitelistedIssuanceModule(address _issuanceModule) {
+        require(allowedIssuanceModules[_issuanceModule].isAllowed, "ExchangeIssuance: INVALID ISSUANCE MODULE");
+         _;
+    }
+
+    constructor(
+        address _weth,
+        IController _setController,
+        address[] memory _issuanceModuleAddresses,
+        bool[] memory _issuanceModuleDebtIssuanceFlags,
+        address _swapTarget
+    )
+        public
+    {
+        require(_issuanceModuleAddresses.length == _issuanceModuleDebtIssuanceFlags.length, "ExchangeIssuance: ISSUANCE MODULE ADDRESSES / TYPE FLAGS LENGTH MISMATCH");
+        setController = _setController;
+
+        WETH = _weth;
+        swapTarget = _swapTarget;
+
+        for (uint256 i = 0; i < _issuanceModuleAddresses.length; i++) {
+            _addIssuanceModule(_issuanceModuleAddresses[i], _issuanceModuleDebtIssuanceFlags[i]);
+        }
+
+    }
+
+    /* ============ External Functions ============ */
+
+    /**
+     * Withdraw slippage to selected address
+     *
+     * @param _tokens    Addresses of tokens to withdraw, specifiy ETH_ADDRESS to withdraw ETH
+     * @param _to        Address to send the tokens to
+     */
+    function withdrawTokens(IERC20[] calldata _tokens, address payable _to) external onlyOwner payable {
+        for(uint256 i = 0; i < _tokens.length; i++) {
+            if(address(_tokens[i]) == ETH_ADDRESS){
+                _to.sendValue(address(this).balance);
+            }
+            else{
+                _tokens[i].safeTransfer(_to, _tokens[i].balanceOf(address(this)));
+            }
+        }
+    }
+
+    receive() external payable {
+        // required for weth.withdraw() to work properly
```

Comment:
How does `receive` get called as part of weth.withdraw? Is this actually about weth.deposit failing?

Reply:
We call `weth.withdraw` when we want to return ETH to the user.

This is done since for ETH trades (i.e. when using ETH as input / output taken during issue/redeem) we actually send a WETH based trade to the 0x Proxy / Router. To return ETH instead of WETH to the user in those cases we call `withdraw` which in turn sends eth to this contract triggering the `receive`.

Reply:
Withdrawing eth to return leftover from issuance:
https://github.com/FlattestWhite/index-coop-smart-contracts/blob/exchange-issuance-zero-ex/contracts/exchangeIssuance/ExchangeIssuanceZeroEx.sol#L274

Withdrawing eth to send out resulting funds from redemption:
https://github.com/FlattestWhite/index-coop-smart-contracts/blob/exchange-issuance-zero-ex/contracts/exchangeIssuance/ExchangeIssuanceZeroEx.sol#L350

Reply:
@cgewecke Does that answer your question ?

Reply:
Yes, thanks so much for this explanation :100: